src/main/java/com/squareup/javapoet/ParameterizedTypeName.java

```diff
@@ -37,37 +42,101 @@
 
   ParameterizedTypeName(ClassName rawType, List<TypeName> typeArguments,
       List<AnnotationSpec> annotations) {
+    this(
+        rawType,
+        rawType.enclosingClassName() != null ? rawType.enclosingClassName() : rawType,
```

Comment:
Could you just change this to be the enclosing type only, and we can use rawType if that’s null?